pennylane/data/dataset.py

```diff
@@ -0,0 +1,157 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Contains the :class:`~pennylane.data.Dataset` class and its associated functions.
+"""
+
+from abc import ABC
+import dill
+import zstd
+
+
+class Dataset(ABC):
+    """The dataset data type. This class is used to create general datasets.
+    It accepts general keyword arguments to be flexible for different types of data.
+
+    Args:
+        dtype (string): the type of dataset, e.g. 'qchem' or 'qspin'
+        **kwargs: variable length of keyword arguments containing data to store in the dataset
+
+    .. details::
+        :title: Usage Details
+
+        We can use the :class:`~pennylane.data.Dataset` class to create datasets as follows:
+
+        .. code-block:: python
+
+            import pennylane as qml
+
+            coefficients = [1]*2
+            observables = [qml.PauliZ(wires=i) for i in range(2)]
+            Hamiltonian = qml.Hamiltonian(coefficients, observables)
+
+            ex_dataset = qml.data.Dataset(Hamiltonian = Hamiltonian)
+
+        >>> print(ex_dataset.Hamiltonian)
+          (1) [Z0]
+        + (1) [Z1]
+
+        To save the dataset to a file, we call :func:`Dataset.write()`:
+
+        >>> ex_dataset.write('./path/to/file/ex_dataset.dat')
+
+        We can then retrieve the data using :func:`Dataset.read()`
+
+        >>> retrieved_data = qml.data.Dataset()
+        >>> retrieved_data.read('./path/to/file/ex_dataset.dat')
+        >>> print(retrieved_data.Hamiltonian)
+          (1) [Z0]
+        + (1) [Z1]
+    """
+
+    def __init__(self, dtype=None, **kwargs):
+        self.dtype = dtype
+        self.__doc__ = ""
+        for key, val in kwargs.items():
+            setattr(self, f"{key}", val)
+
+    def __eq__(self, __o):
```

Comment:
Added this to enable straightforward comparison of two dataset objects.

Reply:
Noticed this doesn't work if the dataset objects contain qml.Hamiltonian objects. But that's because qml.Hamiltonian == qml.Hamiltonian returns false even when the hamiltonians are the same.

Reply:
Then the previous one shouldn't work either I guess lol.

Reply:
Just double checking here, do you want to overwrite `__eq__`? From what I recall, defining `__eq__` has consequences for Python object hashing.

Basically, I would only define this if absolutely needed for something